kimchi/src/alphas.rs

```diff
@@ -0,0 +1,276 @@
+
+/// This type can be used to create a mapping between powers of alpha and constraint types.
+/// See [Builder::default] to create one,
+/// and [Builder::register] to register a new mapping.
+/// Once you know the alpha value, you can convert this type to a [Alphas].
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+pub struct Builder {
+    /// The next power of alpha to use
+    next_power: usize,
+    /// The mapping between constraint types and powers of alpha
+    mapping: HashMap<ConstraintType, Range<usize>>,
+}
+
+impl Builder {
+    /// Registers a new [ConstraintType],
+    /// associating it a number `powers` of powers of alpha.
+    /// The function returns an iterator of powers that must be used.
+    #[must_use]
+    pub fn register(&mut self, ty: ConstraintType, powers: usize) -> Range<usize> {
+        let new_power = self.next_power + powers;
+        let range = self.next_power..new_power;
+        if self.mapping.insert(ty.clone(), range.clone()).is_some() {
+            panic!("you are attempting to register {:?} twice", ty);
+        }
+        self.next_power = new_power;
+        range
+    }
+
+    /// Creates a new instance of [Alphas] via a [Builder] and value `alpha`.
+    pub fn into_alphas<F: Field>(self, alpha: F) -> Alphas<F> {
+        let mut last_power = F::one();
+        let mut alphas = Vec::with_capacity(self.next_power);
+        alphas.push(F::one());
+        for _ in 0..self.next_power {
+            last_power *= alpha;
+            alphas.push(last_power);
+        }
+
+        Alphas {
+            alphas,
+            mapping: self.mapping,
+        }
+    }
+}
+
+// ------------------------------------------
+
+/// This type can be constructed from [Builder] and a value `alpha`,
+/// via [Alphas::new].
+/// It will then pre-compute the necessary numbers of powers of alpha.
+/// You can retrieve powers of alpha by calling [Alphas::take_alphas],
+/// or the exponents without the actual powers of alphas via [Alphas::take_powers].
+pub struct Alphas<F> {
+    /// The powers of alpha: 1, alpha, alpha^2, etc.
+    alphas: Vec<F>,
+    /// The mapping from constraint type to a range of powers of alpha
+    mapping: HashMap<ConstraintType, Range<usize>>,
+}
+
+impl<F: Field> Alphas<F> {
+    /// Creates a new instance of [Alphas] via a [Builder] and value `alpha`.
+    #[must_use]
+    pub fn new(alpha: F, powers: &Builder) -> Alphas<F> {
+        let mut last_power = F::one();
+        let mut alphas = Vec::with_capacity(powers.next_power);
+        alphas.push(F::one());
+        for _ in 0..powers.next_power {
+            last_power *= alpha;
+            alphas.push(last_power);
+        }
+
+        Alphas {
+            alphas,
+            mapping: powers.mapping.clone(),
+        }
+    }
+
+    /// This returns a range of powers (the exponents), upperbounded by `num`
+    #[must_use]
+    pub fn get_powers(
+        &mut self,
+        ty: ConstraintType,
```

Comment:
It's actually really annoying, because both `ty` and `type` are commonly used in code that deals with types.  This is because `type` is often a restricted keyword. I actually like `type_` better.

Reply:
Ok nice, in any case this was just a triviality :)